Modules/_functoolsmodule.c

```diff
@@ -267,12 +416,31 @@ partial_vectorcall(partialobject *pto, PyObject *const *args,
         }
     }
 
-    /* Copy to new stack, using borrowed references */
-    memcpy(stack, pto_args, pto_nargs * sizeof(PyObject*));
-    memcpy(stack + pto_nargs, args, nargs_total * sizeof(PyObject*));
-
-    ret = _PyObject_VectorcallTstate(tstate, pto->fn,
-                                     stack, pto_nargs + nargs, kwnames);
+    Py_ssize_t nargs_new;
+    if (np) {
+        nargs_new = pto_nargs + nargs - np;
+        Py_ssize_t j = 0;       // New args index
+        for (Py_ssize_t i=0; i < pto_nargs; i++) {
+            if (Py_Is(pto_args[i], pto->placeholder)){
+                memcpy(stack + i, args + j, 1 * sizeof(PyObject*));
+                j += 1;
+            }
+            else {
+                memcpy(stack + i, pto_args + i, 1 * sizeof(PyObject*));
+            }
```

Comment:
Instead of calling multiple `memcpy`, I'd suggest picking up the longest window of placeholder-like objects and then use a single `memcpy` over a range.

Reply:
Same point here. Time saving would be small, if any, but it would be more complex.
E.g.:
```python
from partial2 import partial, Placeholder
f = lambda a, b, c, d, e, f: (a, b, c, d, e, f)
fp = ftl.partial(f, 0, 1, 2)
pp = partial(f, PH, 1, PH, 3, PH, 5)

In [11]: fp2(3, 4, 5)
Out[11]: (0, 1, 2, 3, 4, 5)
In [12]: pp2(0, 2, 4)
Out[12]: (0, 1, 2, 3, 4, 5)

In [17]: %timeit fp2(3, 4, 5)
143 ns ± 2.75 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
In [18]: %timeit pp2(3, 4, 5)
148 ns ± 0.442 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Will keep this in mind as I go forward, but for now extra complexity doesn't seem to be worth it.